### PR TITLE
[Core][Migration] Support waiting request and multiple requests migration

### PR DESCRIPTION
**Pull request by s5u13b**
Previously, Llumnix only support migrate running request and migrate one request once at most. In this PR, we support migrate waiting request and migrate multiple requests one by one.
Besides, 
1. Add 'FCW' and 'FCWSR' request migration policy, the detailed explanation of policy can be found in arg_utils.py.
2. Add explanations of policy arguments in arguments' help option. 
3. Rename request migration policy to make them more clear.
4. Slightly refine some migration implementation (free_dst_pre_alloc_cache etc).
5. Fix some bugs in unit test (migration and pd disaggregation).
6. Fix bug in pd disaggregation (expected_steps).

### CODE REVIEW

File: llumnix/backends/vllm/scheduler.py [hunk 1]
```diff
@@ -110,28 +126,63 @@ def pop_migrating_out_requests_last_stage(self) -> List[SequenceGroupLlumnix]:
...
+                  request_id: str,
+                  request_status: RequestStatus,
+                  request_arrival_time: float,
+                  block_num: int) -> List[int]:
+        # Only migrate waiting request when the waiting request is the earliest arrival one
+        # among the requests of dst instance's waiting queue.
+        if request_status == RequestStatus.WAITING_MIGRATING:
+            if (self.waiting and request_arrival_time > self.waiting[0].arrival_time) \
+                or block_num * self.cache_config.block_size > self.prompt_limit:
+                return []
```
Comment (zhypku) [inline llumnix/backends/vllm/scheduler.py:139]: Can you return None?

Comment (s5u13b) [reply llumnix/backends/vllm/scheduler.py:139]: OK.

Comment (zhypku) [reply llumnix/backends/vllm/scheduler.py:139]: I'm the only non-bot commenter

Comment (s5u13b) [reply llumnix/backends/vllm/scheduler.py:139]: `return []` is also a common coding style, vllm also applies the `return []` way, so i think we can stay unchanged here.

File: llumnix/llumlet/local_migration_scheduler.py [hunk 2]
```diff
@@ -11,63 +11,92 @@
...
-        condition = lambda request : request.inference_type == RequestInferenceType.DECODE \
-                                        and min_request_len <= request.request_len <= max_request_len
+    def _filter_waiting_queue(self, waiting, min_request_len, max_request_len):
+        filtered_waiting = [
+            request for request in waiting \
+                if request.status == RequestStatus.WAITING \
+                    and request.try_schedule_times >= 1 \
+                    and min_request_len < request.request_len < max_request_len \
+        ]
+        return filtered_waiting
```
Comment (KuilongCui) [inline llumnix/llumlet/local_migration_scheduler.py:71]: I think there is no need to use try_schedule_times at this stage. For now,  we should keep our approach overall concise and defer any complex design of strategies to a later step.

Comment (KuilongCui) [reply llumnix/llumlet/local_migration_scheduler.py:71]: Using min_request_len and max_request_len to determine the migration of multiple requests lacks flexibility and scalability. I suggest separating the migration of multiple requests and the migration of waiting requests into different pull requests (PRs).

File: llumnix/llumlet/migration_coordinator.py [hunk 3]
```diff
@@ -50,65 +48,110 @@ def __init__(self,
         self.max_stages = max_stages
         self.backend_engine = backend_engine
 
-    async def migrate_out_onestage(self, migrate_in_ray_actor: "ray.actor.ActorHandle",  migrate_out_request: LlumnixRequest, ) -> "MigrationStatus":
-        """one-stage live migration until last stage
+    async def migrate_out_running_request(self,
```
Comment (KuilongCui) [inline llumnix/llumlet/migration_coordinator.py:51]: Is it possible to integrate the logic for migrating running and waiting requests> In the future, we may also need to migrate swapped requests, so maintaining overall indirectness and extensibility is important.

File: llumnix/backends/backend_interface.py [hunk 4]
```diff
@@ -99,14 +99,21 @@ def get_request_incremental_blocks(self, backend_request: LlumnixRequest, pre_st
         raise NotImplementedError
 
     @abstractmethod
-    def get_running_queue(self) -> List[LlumnixRequest]:
+    def get_running_queue(self) -> Deque[LlumnixRequest]:
```
[the remark this answers is not in the document]
Comment (s5u13b) [reply llumnix/backends/backend_interface.py:102]: The request of vllm is the deque type.

File: llumnix/backends/vllm/scheduler.py [hunk 5]
```diff
@@ -141,7 +192,8 @@ def free_dst_pre_alloc_cache(self, request_id: str = None) -> None:
 
     def free_src_request(self, backend_request: SequenceGroupLlumnix) -> None:
         seq = backend_request.get_seqs()[0]
-        logger.info("free seq {}".format(seq.seq_id))
+        logger.info("free request: {}".format(backend_request.request_id))
+        logger.info("free seq: {}".format(seq.seq_id))
```
[the remark this answers is not in the document]
Comment (s5u13b) [reply llumnix/backends/vllm/scheduler.py:196]: OK.

File: llumnix/backends/backend_interface.py [hunk 6]
```diff
@@ -164,17 +191,25 @@ def pop_migrating_out_requests_last_stage(self) -> List[LlumnixRequest]:
         raise NotImplementedError
 
     @abstractmethod
-    def pre_alloc(self, request_id: str, block_num: int) -> List[int]:
+    def pre_alloc(self,
+                  request_id: str,
+                  request_status: RequestStatus,
+                  request_arrival_time: float,
+                  block_num: int) -> List[int]:
```
[the remark this answers is not in the document]
Comment (s5u13b) [reply llumnix/backends/backend_interface.py:198]: Only the request whose arrival time is earliest among the waiting requests of dst instance could be migrated in. I agree with you but i can't find better way to implement it. The pre_alloc function now acts as can_pre_alloc plus pre_alloc in fact.

File: llumnix/backends/vllm/llm_engine.py [hunk 7]
```diff
@@ -350,15 +352,22 @@ def commit_dst_request(self, backend_request: SequenceGroupLlumnix) -> None:
...
-        self.add_running_request(backend_request)
+        backend_request.reset_migration_args_dst()
+        assert backend_request.status in [RequestStatus.WAITING_MIGRATING, RequestStatus.RUNNING_MIGRATING], \
+            "The status of request migrated to dst instance should be  \
+             RequestStatus.WAITING_MIGRATING or RequestStatus.RUNNING_MIGRATING"
+        if backend_request.status == RequestStatus.WAITING_MIGRATING:
+            self.add_waiting_request(backend_request)
+        elif backend_request.status == RequestStatus.RUNNING_MIGRATING:
+            backend_request.reset_status()
+            self.add_running_request(backend_request)
```
[the remark this answers is not in the document]
Comment (s5u13b) [reply llumnix/backends/vllm/llm_engine.py:363]: OK.

File: llumnix/llumlet/local_migration_scheduler.py (hunk 2 above)
Comment (s5u13b) [reply llumnix/llumlet/local_migration_scheduler.py:71]: The concept of min_request_len/max_request_len had been introduced previously. It is introduced for the reason that only request with proper length can achieve relatively good migration bandwidth and scheduling result. So I think we should keep it temporarily for the integrity of design.

File: llumnix/llumlet/migration_coordinator.py (hunk 3 above)
Comment (s5u13b) [reply llumnix/llumlet/migration_coordinator.py:51]: Migrating running requests and waiting requests have different stages of migration, the running requests are migrated in multiple stages and the waiting requests are migrated in one stage, and there are other slight differences also. It is hard to integrate them together in a elegant manner. So I think there is no need to over-design currently.

**Review by zhypku — APPROVED** [review]

**Review by KuilongCui — APPROVED** [review]

File: tests/e2e_test/test_migration.py [hunk 8]
```diff
@@ -99,23 +105,23 @@ async def run_bench_command(command):
 
     _, pending = await asyncio.wait(tasks, timeout=60*30)
```
[the remark this answers is not in the document]
Comment (KuilongCui) [reply tests/e2e_test/test_migration.py:106]: and bench_test

**Pull request merged**